### PR TITLE
Some more DAO improvements

### LINKED ISSUE #17: Add JSON persistence of a User object
This requires a new data access class, which writes to and reads from a JSON file. Will need a `load` and `save` method, and perhaps some other helper methods.

The class will need some way to create users, some way to access the Recipe API, a method of creating recipes, and the file path that we will read to / write from.

### PR DESCRIPTION
**Pull request by pratzyushd**
Implement the JSON persistence functionality (this is functional, but still requires modification to be optimal).

Add test cases for JSON persistence.

Also some mild refactoring of the interface for the Recipe DB API.

Once completed and merged, this should close #17.

### CODE REVIEW

File: src/data_access/JSONPersistence.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                array.put(recipe.getMealId());
+            }
+            /* Add the tag as the key and the array as the values. */
+            obj.put(tag, array);
+        }
+        return obj;
+    }
+
+    /**
+     * Create a JSONArray form the given User object's favourites.
```
Comment (jaeyonglee3) [inline src/data_access/JSONPersistence.java:56]: Typo in the Javadocs here, this should say "Create a JSONArray **_from_** the given User object's favourites."

File: src/data_access/JSONPersistence.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    /**
+     * Make a JSONObject out of the tags for a given user.
+     * @param user the User object which we make the JSONObject based on.
+     * @return JSONObject containing a "tags" key and then an array of
+     * JSONObjects, each of which has a key of the tag name, with an associated
+     * JSONArray for the recipe IDs.
+     */
+    private JSONObject JSONForUserTags(User user) {
+        JSONObject obj = new JSONObject();
+        JSONArray array = new JSONArray();
```
Comment (jaeyonglee3) [inline src/data_access/JSONPersistence.java:39]: The JSONArray should be created within the loop for each tag to avoid accumulating recipe IDs for all tags in a single array and unintentionally duplicating recipes across multiple tags.

File: src/data_access/JSONPersistence.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        try {
+            writer = new BufferedWriter(new FileWriter(this.filePath));
+            writer.write(obj.toString(4));
+            writer.close();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public User load() {
```
Comment (jaeyonglee3) [inline src/data_access/JSONPersistence.java:107]: Missing Javadocs for load() method.

File: src/data_access/JSONPersistence.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+            JSONObject obj = new JSONObject(tokenizer);
+            name = obj.getString("username");
+            favourites = obj.getJSONArray("favourites");
+            tagsMap = obj.getJSONObject("tags");
+        } catch (JSONException | IOException e) {
+            throw new RuntimeException(e);
+        }
+        /* Construct the user object. */
+        /* TODO: fix this hack, where we don't use the password. */
+        user = userFactory.create(name, "");
```
Comment (jaeyonglee3) [inline src/data_access/JSONPersistence.java:125]: As a note, the change to the User Factory file to remove the password parameter from its create method has been taken care of and pushed to the main branch.

File: src/data_access/JSONPersistence.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        /* Update user with favourites */
+        Recipe[] favouritesList = generateRecipesFromJSONArray(favourites);
+        for (Recipe recipe : favouritesList) {
+            user.addFavourite(recipe);
```
Comment (jaeyonglee3) [inline src/data_access/JSONPersistence.java:129]: As a note for the future, for the sake of efficiency when loading users, it may be useful to define a 'setFavourites' method in the User class so we don't have to iterate through each recipe and add them one by one. We could also overload the existing addFavourite method to also accept an entire array of recipes.

File: src/data_access/JSONPersistence.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        }
+        /* Update user with tags */
+        for (String key: tagsMap.keySet()) {
```
Comment (jaeyonglee3) [inline src/data_access/JSONPersistence.java:132]: Again, same thing as my comment above on line 129, it may be beneficial to overload the 'assignTag' method in the User class and allow it to accept an entire hash map as well to avoid this need to iterate through tagsMap.

File: src/data_access/JSONPersistence.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+            Recipe[] recipeArray = generateRecipesFromJSONArray(tagsMap.getJSONArray(key));
+            for (Recipe recipe : recipeArray) {
+                user.assignTag(recipe, key);
+            }
+        }
+        return user;
+    }
+
+    private Recipe[] generateRecipesFromJSONArray(JSONArray array) {
```
Comment (jaeyonglee3) [inline src/data_access/JSONPersistence.java:141]: A more efficient way to query a list of IDs and get back a list of Recipe objects is a great next step, perhaps something we could work on together because this seems interesting to me.

File: src/data_access/Persistence.java [hunk 8]
```diff
@@ -0,0 +1,6 @@
+package data_access;
+
+/* TODO: this needs to be modified to include the interfaces from the other
+ * use cases. */
+public interface Persistence {
```
Comment (jaeyonglee3) [inline src/data_access/Persistence.java:5]: I'm not totally clear on the purpose of this Persistence interface, if some clarifying comments or Javadocs could be added that would be great.

File: src/data_access/TheMealDB.java [hunk 9]
```diff
@@ -27,7 +27,7 @@ public TheMealDB(RecipeFactory recipeFactory) {
         this.recipeFactory = recipeFactory;
     }
     @Override
-    public Recipe getRecipeById(String id) {
+    public Recipe searchRecipesById(String id) {
```
Comment (jaeyonglee3) [inline src/data_access/TheMealDB.java:30]: Small nitpick here but this probably should be named "searchRecipeById" since we are only expecting 1 recipe back from the API. This can be refactored later as I'm assuming other files are already making use of this.

File: test/data_access/JSONPersistenceTest.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    @Test
+    public void testEmptyUserOutput() throws IOException {
+        instance.save(emptyUser);
+        String fileContents = new String(Files.readAllBytes(Path.of(this.filePath)));
+        assertTrue(fileContents.contains("favourites"));
+        assertTrue(fileContents.contains("tags"));
+        assertTrue(fileContents.contains("{}"));
+        assertTrue(fileContents.contains("[]"));
+    }
+    @Test
```
Comment (jaeyonglee3) [inline test/data_access/JSONPersistenceTest.java:69]: Add a line break here.

File: test/data_access/JSONPersistenceTest.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        ArrayList<Recipe> taggedRecipes = user.getTaggedRecipes().get("tag");
+        for (int i = 0; i < taggedRecipes.size(); i++) {
+            taggedIds[i] = taggedRecipes.get(i).getMealId();
+        }
+
+        assertEquals(favouriteIds.length, 2);
+        assertTrue(actualFavouriteIds.containsAll(List.of(favouriteIds)));
+        assertTrue(actualTaggedIds.containsAll(List.of(taggedIds)));
+    }
+}
```
Comment (jaeyonglee3) [inline test/data_access/JSONPersistenceTest.java:121]: Another test to add here would be a test that ensures the JSONException is thrown if we try to read an empty JSON file or a JSON file that is not formatted in the way in which we expect.

**Review by jaeyonglee3 — CHANGES_REQUESTED** [review]
Overall amazing stuff as per usual. The consistent use of the enhanced for-loop ('for-each' loop) and detailed Javadocs throughout made the code easily readable. The tests are quite comprehensive already as well. I've added comments throughout with notes and suggestions for moving forward, as well as changes that should be implemented before this PR is merged.

Lastly, the temp file in the test/data_access directory should be deleted.

File: src/data_access/TheMealDB.java (hunk 9 above)
Comment (pratzyushd) [reply src/data_access/TheMealDB.java:30]: Yeah, I noticed this too, but this name is derived from another interface (specifically the `SearchIdUserDataAccessInterface`).

File: src/data_access/Persistence.java (hunk 8 above)
Comment (jaeyonglee3) [reply src/data_access/Persistence.java:12]: Thank you for adding Javadocs!

**Review by jaeyonglee3 — APPROVED** [review]
LGTM! Looks like you addressed all of my points in the first review. Thank you for your diligence and hard work!
